src/main/java/org/folio/services/financedata/FinanceDataService.java

```diff
@@ -1,39 +1,206 @@
     this.restClient = restClient;
     this.acqUnitsService = acqUnitsService;
+    this.transactionApiService = transactionApiService;
+    this.fiscalYearService = fiscalYearService;
+    this.fundUpdateLogService = fundUpdateLogService;
   }
 
   public Future<FyFinanceDataCollection> getFinanceDataWithAcqUnitsRestriction(String query, int offset, int limit,
                                                                                RequestContext requestContext) {
+    log.debug("Trying to get finance data with acq units restriction, query={}", query);
     return acqUnitsService.buildAcqUnitsCqlClauseForFinanceData(requestContext)
       .map(clause -> StringUtils.isEmpty(query) ? clause : combineCqlExpressions("and", clause, query))
       .compose(effectiveQuery -> getFinanceData(effectiveQuery, offset, limit, requestContext));
   }
 
-  private Future<FyFinanceDataCollection> getFinanceData(String query, int offset, int limit, RequestContext requestContext) {
+  private Future<FyFinanceDataCollection> getFinanceData(String query, int offset, int limit,
+                                                         RequestContext requestContext) {
     var requestEntry = new RequestEntry(resourcesPath(FINANCE_DATA_STORAGE))
       .withOffset(offset)
       .withLimit(limit)
       .withQuery(query);
     return restClient.get(requestEntry.buildEndpoint(), FyFinanceDataCollection.class, requestContext);
   }
+
+  public Future<Void> putFinanceData(FyFinanceDataCollection financeDataCollection, RequestContext requestContext) {
+    log.debug("Trying to update finance data collection with size: {}", financeDataCollection.getTotalRecords());
+    if (CollectionUtils.isEmpty(financeDataCollection.getFyFinanceData())) {
+      log.info("Finance data collection is empty, nothing to update");
+      return succeededFuture();
+    }
+
+    validateFinanceDataCollection(financeDataCollection, getFiscalYearId(financeDataCollection));
+
+    return processAllocationTransaction(financeDataCollection, requestContext)
+      .compose(v -> updateFinanceData(financeDataCollection, requestContext))
+      .onSuccess(asyncResult -> processLogs(financeDataCollection, requestContext, COMPLETED))
+      .onFailure(asyncResult -> processLogs(financeDataCollection, requestContext, ERROR));
+  }
+
+  private void validateFinanceDataCollection(FyFinanceDataCollection financeDataCollection, String fiscalYearId) {
+    for (int i = 0; i < financeDataCollection.getFyFinanceData().size(); i++) {
+      var financeData = financeDataCollection.getFyFinanceData().get(i);
+      validateFinanceDataFields(financeData, i, fiscalYearId);
+
+      var allocationChange = BigDecimal.valueOf(financeData.getBudgetAllocationChange());
+      var initialAllocation = BigDecimal.valueOf(financeData.getBudgetInitialAllocation());
+
+      if (allocationChange.doubleValue() < 0 && allocationChange.abs().doubleValue() > initialAllocation.doubleValue()) {
```

Comment:
Why not use only Doubles or BigDecimals here?

Reply:
There was some calculation previously. I forgot to revert back them. changed to double